### PR TITLE
[WIP] Integrate coinbase puzzle into ledger

### PR DESCRIPTION
**Pull request by raychu86**
## Motivation

Based on https://github.com/AleoHQ/snarkVM/pull/994

This PR cleans up the new posw coinbase puzzle and begins to integrate its use into the ledger.

### CODE REVIEW

File: vm/compiler/benches/coinbase_puzzle/coinbase_puzzle.rs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+#[macro_use]
+extern crate criterion;
+
+use console::{account::*, network::Testnet3};
+use snarkvm_compiler::{CoinbasePuzzle, EpochChallenge, EpochInfo};
+
+use criterion::Criterion;
+use rand::{self, thread_rng, CryptoRng, RngCore};
```
Comment (ljedrz) [inline vm/compiler/benches/coinbase_puzzle/coinbase_puzzle.rs:24]: when https://github.com/AleoHQ/snarkVM/pull/1013 is merged, the slow and non-deterministic `thread_rng` could be replaced with `TestRng` here

File: vm/compiler/src/coinbase_puzzle/hash.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+use snarkvm_fields::PrimeField;
+use snarkvm_utilities::{cfg_into_iter, CanonicalSerialize};
+
+use anyhow::Result;
+use blake2::Digest;
+
+#[cfg(feature = "parallel")]
+use rayon::prelude::*;
+
+pub fn hash_to_poly<F: PrimeField>(input: &[u8], degree: usize) -> Result<DensePolynomial<F>> {
```
Comment (ljedrz) [inline vm/compiler/src/coinbase_puzzle/hash.rs:28]: shouldn't degree be `u64` explicitly? it seems like the size of `input_with_counter` depends on it being 8 bytes in size

File: vm/compiler/src/coinbase_puzzle/hash.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            input_with_counter[..32].copy_from_slice(&input_hash);
+            input_with_counter[32..].copy_from_slice(&i.to_le_bytes());
+            F::from_bytes_le_mod_order(&blake2::Blake2b512::digest(&input_with_counter))
+        })
+        .collect::<Vec<_>>();
+    Ok(DensePolynomial::from_coefficients_vec(coefficients))
+}
+
+pub fn hash_commitment<E: PairingEngine>(cm: &Commitment<E>) -> E::Fr {
+    let mut bytes = Vec::with_capacity(48);
```
Comment (ljedrz) [inline vm/compiler/src/coinbase_puzzle/hash.rs]:
```suggestion
    let mut bytes = [0u8; 48];
```

File: vm/compiler/src/coinbase_puzzle/hash.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    // This is faster than compressed serialization that's provided by `ToBytes`.
+    cm.serialize_uncompressed(&mut bytes).unwrap();
+    E::Fr::from_bytes_le_mod_order(&blake2::Blake2b512::digest(&bytes))
+}
+
+pub fn hash_commitments<E: PairingEngine>(cms: impl ExactSizeIterator<Item = Commitment<E>>) -> Vec<E::Fr> {
+    let num_commitments = cms.len();
+    let cm_bytes = cms
+        .flat_map(|c| {
+            let mut bytes = Vec::with_capacity(48);
```
Comment (ljedrz) [inline vm/compiler/src/coinbase_puzzle/hash.rs]: I wonder if this shouldn't be `[0u8; 48]` instead too for better performance; unless there can be a _lot_ of them

File: vm/compiler/src/coinbase_puzzle/hash.rs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+pub fn hash_commitment<E: PairingEngine>(cm: &Commitment<E>) -> E::Fr {
+    let mut bytes = Vec::with_capacity(48);
+    // This is faster than compressed serialization that's provided by `ToBytes`.
+    cm.serialize_uncompressed(&mut bytes).unwrap();
+    E::Fr::from_bytes_le_mod_order(&blake2::Blake2b512::digest(&bytes))
+}
+
+pub fn hash_commitments<E: PairingEngine>(cms: impl ExactSizeIterator<Item = Commitment<E>>) -> Vec<E::Fr> {
+    let num_commitments = cms.len();
```
Comment (ljedrz) [inline vm/compiler/src/coinbase_puzzle/hash.rs]:
```suggestion
    let num_commitments = cms.len() as u64;
```
(same reasoning as before)

File: vm/compiler/src/coinbase_puzzle/tests.rs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+use super::*;
+use console::{account::*, network::Testnet3};
+use snarkvm_utilities::Uniform;
+
+use rand::RngCore;
+
+#[test]
+fn test_coinbase_puzzle() {
+    let max_degree = 1 << 15;
+    let mut rng = rand::thread_rng();
```
Comment (ljedrz) [inline vm/compiler/src/coinbase_puzzle/tests.rs]:
```suggestion
    let mut rng = snarkvm_utilities::test_rng();
```
or `TestRng::default()` if https://github.com/AleoHQ/snarkVM/pull/1013 is merged first

File: vm/compiler/src/ledger/mod.rs [hunk 7]
```diff
@@ -315,12 +324,29 @@ impl<N: Network, B: BlockStorage<N>, P: ProgramStorage<N>> Ledger<N, B, P> {
         Ok(())
     }
 
+    /// Appends the given transaction to the memory pool.
+    pub fn add_to_prover_puzzle_memory_pool(&mut self, prover_puzzle_solution: ProverPuzzleSolution<N>) -> Result<()> {
+        // Ensure the transaction does not already exist.
+        if self.prover_puzzle_memory_pool.contains(&prover_puzzle_solution) {
```
Comment (ljedrz) [inline vm/compiler/src/ledger/mod.rs]: once the `prover_puzzle_memory_pool` is an `IndexMap`, the `contains` will no longer be needed, as long as the value is the same or it's an `IndexSet` (as `insert` will indicate if the key had already existed)

**Comment by HarukaMa** [comment]
I'm getting this when trying the coinbase puzzle benchmark (`$ RUSTFLAGS="-C target-cpu=native" cargo bench --bench coinbase_puzzle`):

```Benchmarking CoinbasePuzzle::Prove 32: Warming up for 3.0000 sthread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: the number of coefficients in the polynomial (62) is greater thanthe maximum number of powers in `Powers` (33)', vm/compiler/benches/coinbase_puzzle/coinbase_puzzle.rs:75:102```

**Comment by ljedrz** [comment]
@HarukaMa wasn't the Rng seed printed out as well?

edit: ah, no, this PR still uses `thread_rng` as opposed to `TestRng`, so it wouldn't be logged

File: vm/compiler/benches/coinbase_puzzle/coinbase_puzzle.rs (hunk 1 above)
Comment (ljedrz) [reply vm/compiler/benches/coinbase_puzzle/coinbase_puzzle.rs:24]: #1013 was merged, so `TestRng` can be used now instead; it implements `CryptoRng` too, so it's compatible

**Comment by raychu86** [comment]
Closing in favor of #1100

**Pull request closed without merging**